paddle/framework/tensor.h

```diff
@@ -164,6 +170,80 @@ class Tensor {
   size_t offset_;
 };
 
+/*
+ * LODTensor
+ */
+class LODTensor {
+ public:
+// level_t save offsets of each unit.
+#ifdef PADDLE_ONLY_CPU
+  using level_t = std::vector<uint32_t>;
+#else
+  using level_t = thrust::device_vector<uint32_t>;
+#endif
+  // LOD stores offsets of each level of units, the largest units level first,
+  // then the smaller units level. Each level_t stores the offsets of units in
+  // Tesor.
+  typedef std::vector<level_t> lod_t;
+
+  explicit LODTensor() {}
+
+  LODTensor(LODTensor&& other)
+      : lod_start_pos_(std::move(other.lod_start_pos_)),
+        tensor_(other.tensor_) {}
+
+  /*
+   * Number of LODTensor's levels, each level has units of data, for example,
+   * in the sentence's view, article, paragraph, sentence are 3 levels.
+   */
+  size_t Levels() const {
+    return lod_start_pos_.get() ? lod_start_pos_->size() : 0UL;
+  }
+  /*
+   * Number of elements in a level.
+   */
+  size_t Elements(uint32_t level = 0) const {
+    PADDLE_ENFORCE(level < Levels(), "level [%d] out of range [%d]", level,
+                   Levels());
+    return lod_start_pos_->at(level).size();
+  }
+
+  /*
+   * Slice of a level.
+   */
+  LODTensor Slice(uint32_t level) const;
+
+  /*
+   * Slice of elements of a level, [elem_begin: elem_end]
+   * NOTE low performance in slice seq_start_positions_.
+   */
+  template <typename T>
+  LODTensor Slice(uint32_t level, uint32_t elem_begin, uint32_t elem_end) const;
+
+  // Slice with tensor's data shared with this.
+  // LODTensor LODSliceShared(int level, int elem_begin, int elem_end) const;
+
+  // Copy other's lod_start_pos_, to share LOD info.
+  // NOTE the LOD info sould not be changed.
+  void ShareConstLODFrom(const LODTensor& other) {
+    lod_start_pos_ = other.lod_start_pos_;
+  }
+  // Copy other's lod_start_pos_'s content, free to mutate.
+  void ShareMutableLODFrom(const LODTensor& other) {
+    lod_start_pos_ = std::make_shared<lod_t>(*other.lod_start_pos_);
+  }
+  // Determine whether LODTensor has a valid LOD info.
+  bool HasLOD() const { return lod_start_pos_.get(); }
+
+  void set_tensor(std::shared_ptr<Tensor> tensor) { tensor_ = tensor; }
+  void set_lod(std::shared_ptr<lod_t> lod) { this->lod_start_pos_ = lod; }
```

Comment:
Make `set_lod`/`HasLOD` name consistently. It is confused for user who will use this class.